File: android/app/src/main/kotlin/dev/imranr/obtainium/FleetProfileApplier.kt
package dev.imranr.obtainium

import android.content.Context
import android.net.Uri
import androidx.datastore.preferences.core.MutablePreferences
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.runBlocking
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Must match SharedPreferencesPlugin.kt's DataStore name so the native
 * applier writes to the same store that Flutter's shared_preferences reads.
 */
private val Context.fleetPreferencesDataStore by
    preferencesDataStore("FlutterSharedPreferences")

/**
 * Applies a fleet/headless configuration profile to Obtainium's DataStore.
 *
 * Flutter's shared_preferences_android 2.4.26 plugin uses Jetpack DataStore
 * (preferencesDataStore("FlutterSharedPreferences")) as its primary backend,
 * NOT XML SharedPreferences. This applier writes to the same DataStore via
 * the same extension property ([sharedPreferencesDataStore]) that the Flutter
 * plugin uses, so settings are immediately visible to the Dart side.
 *
 * Handles simple preferences (bool, int/long, string). For complex settings
 * (List<String> with JSON_LIST_PREFIX encoding, app imports, update triggers)
 * use the Flutter deep-link route (obtainium://profile?profile=...).
 *
 * Example profile:
 * {
 *   "_meta": { "name": "My profile", "version": 1 },
 *   "installMethod": "shizuku",
 *   "enableBackgroundUpdates": true,
 *   "bgUpdatesOnWiFiOnly": true,
 *   "updateInterval": 360,
 *   "theme": "dark"
 * }
 */
object FleetProfileApplier {

    private val aliasToKey: Map<String, String> = mapOf(
        "installMethod" to "installMethod",
        "installerMode" to "installMethod",
        "externalInstallerPackage" to "externalInstallerPackage",
        "externalInstallerComponent" to "externalInstallerComponent",
        "useSystemFont" to "useSystemFont",
        "theme" to "theme",
        "themeColor" to "themeColor",
        "colourSchemeMode" to "colourSchemeMode",
        "useBlackTheme" to "useBlackTheme",
        "updateInterval" to "updateInterval",
        "checkOnStart" to "checkOnStart",
        "sortColumn" to "sortColumn",
        "sortOrder" to "sortOrder",
        "showAppWebpage" to "showAppWebpage",
        "pinUpdates" to "pinUpdates",
        "buryNonInstalled" to "buryNonInstalled",
        "groupBy" to "groupBy",
        "hideTrackOnlyWarning" to "hideTrackOnlyWarning",
        "hideAPKOriginWarning" to "hideAPKOriginWarning",
        "showAppDowngradeError" to "showAppDowngradeError",
        "tactileFeedbackEnabled" to "tactileFeedbackEnabled",
        "includePrereleasesByDefault" to "includePrereleasesByDefault",
        "removeOnExternalUninstall" to "removeOnExternalUninstall",
        "checkUpdateOnDetailPage" to "checkUpdateOnDetailPage",
        "disablePageTransitions" to "disablePageTransitions",
        "reversePageTransitions" to "reversePageTransitions",
        "enableBackgroundUpdates" to "enableBackgroundUpdates",
        "bgUpdatesOnWiFiOnly" to "bgUpdatesOnWiFiOnly",
        "bgUpdatesWhileChargingOnly" to "bgUpdatesWhileChargingOnly",
        "highlightTouchTargets" to "highlightTouchTargets",
        "disableSwipeActions" to "disableSwipeActions",
        "alwaysUsePhoneLayout" to "alwaysUsePhoneLayout",
        "autoExportOnChanges" to "autoExportOnChanges",
        "exportSettings" to "exportSettings",
        "onlyCheckInstalledOrTrackOnlyApps" to "onlyCheckInstalledOrTrackOnlyApps",
        "parallelDownloads" to "parallelDownloads",
        "showActionBannerForUpdateOnly" to "showActionBannerForUpdateOnly",
        "beforeNewInstallsShareToAppVerifier" to "beforeNewInstallsShareToAppVerifier",
        "shizukuPretendToBeGooglePlay" to "shizukuPretendToBeGooglePlay",
        "categories" to "categories",
        "forcedLocale" to "forcedLocale",
    )

    private val valueAliasToKey: Map<String, Map<String, Any>> = mapOf(
        "installMethod" to mapOf(
            "system" to "system",
            "shizuku" to "shizuku",
            "external" to "external",
        ),
        "groupBy" to mapOf(
            "none" to "none",
            "category" to "category",
            "source" to "source",
        ),
        "theme" to mapOf(
            "system" to 0,
            "light" to 1,
            "dark" to 2,
        ),
        "colourSchemeMode" to mapOf(
            "standard" to 0,
            "vibrant" to 1,
            "expressive" to 2,
            "materialYou" to 3,
        ),
        "sortColumn" to mapOf(
            "added" to 0,
            "nameAuthor" to 1,
            "authorName" to 2,
            "releaseDate" to 3,
        ),
        "sortOrder" to mapOf(
            "ascending" to 0,
            "descending" to 1,
        ),
        "exportSettings" to mapOf(
            "disabled" to 0,
            "enabled" to 1,
            "overwrite" to 2,
        ),
    )

    data class Result(
        val success: Boolean,
        val appliedCount: Int,
        val skippedCount: Int,
        val errors: List<String>,
        val message: String,
    )

    @JvmStatic
    fun applyJson(context: Context, json: String): Result {
        return try {
            val profile = JSONObject(json)
            applyProfile(context, profile)
        } catch (e: Exception) {
            val msg = e.message ?: "Invalid JSON"
            Result(false, 0, 0, listOf(msg), "Profile parse failed: $msg")
        }
    }

    @JvmStatic
    fun applyFromPath(context: Context, path: String): Result {
        return try {
            val json = File(path).readText(Charsets.UTF_8)
            applyJson(context, json)
        } catch (e: Exception) {
            val msg = e.message ?: "Read error"
            Result(false, 0, 0, listOf(msg), "Failed to read $path: $msg")
        }
    }

    @JvmStatic
    fun applyFromUri(context: Context, uri: Uri): Result {
        return try {
            val stream = context.contentResolver.openInputStream(uri)
                ?: return Result(false, 0, 0, listOf("Cannot open URI"),
                    "Cannot open URI: $uri")
            val json = stream.use { it.reader(Charsets.UTF_8).readText() }
            applyJson(context, json)
        } catch (e: Exception) {
            val msg = e.message ?: "URI error"
            Result(false, 0, 0, listOf(msg), "Failed to read URI $uri: $msg")
        }
    }

    private fun applyProfile(context: Context, profile: JSONObject): Result {
        val errors = mutableListOf<String>()
        val meta = profile.optJSONObject("_meta")
        val clearExisting = meta?.optBoolean("clear_existing", false) ?: false

        var applied = 0
        var skipped = 0

        runBlocking {
            context.fleetPreferencesDataStore.edit { prefs ->
                if (clearExisting) {
                    prefs.clear()
                }

                val keys = profile.keys()
                while (keys.hasNext()) {
                    val rawKey = keys.next()
                    if (rawKey.startsWith("_")) continue

                    val prefKey = resolveKey(rawKey)
                    if (prefKey == null) {
                        skipped++
                        errors.add("Unknown key: $rawKey")
                        continue
                    }

                    try {
                        val value = profile.get(rawKey)
                        val resolvedValue = resolveValue(prefKey, value)
                        putValue(prefs, prefKey, resolvedValue)
                        applied++
                    } catch (e: Exception) {
                        skipped++
                        errors.add("$rawKey: ${e.message}")
                    }
                }
            }
        }

        val message = "Applied $applied preferences, skipped $skipped" +
                if (errors.isEmpty()) "" else " (${errors.size} errors)"

        return Result(errors.isEmpty(), applied, skipped, errors, message)
    }

    private fun resolveKey(rawKey: String): String? = aliasToKey[rawKey]

    private fun resolveValue(prefKey: String, value: Any?): Any? {
        if (value !is String) return value
        return valueAliasToKey[prefKey]?.get(value) ?: value
    }

    private fun putValue(
        prefs: MutablePreferences,
        key: String,
        value: Any?,
    ) {
        when (value) {
            is Boolean -> prefs[booleanPreferencesKey(key)] = value
            is String -> prefs[stringPreferencesKey(key)] = value
            is Int -> prefs[longPreferencesKey(key)] = value.toLong()
            is Long -> prefs[longPreferencesKey(key)] = value
            is Double -> prefs[stringPreferencesKey(key)] =
                DOUBLE_PREFIX + value.toString()
            is Float -> prefs[stringPreferencesKey(key)] =
                DOUBLE_PREFIX + value.toDouble().toString()
            is JSONArray -> prefs[stringPreferencesKey(key)] = value.toString()
            else -> throw IllegalArgumentException(
                "Unsupported type: ${value?.javaClass?.simpleName}")
        }
    }

    /**
     * Must match SharedPreferencesPlugin.kt DOUBLE_PREFIX so that doubles
     * stored by the native applier are decoded correctly by the Dart side.
     */
    private const val DOUBLE_PREFIX = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBEb3VibGUu"
}

File: lib/providers/fleet_profile_applier.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:obtainium/providers/apps_provider.dart';
import 'package:obtainium/providers/logs_provider.dart';
import 'package:obtainium/providers/settings_provider.dart';
import 'package:shizuku_apk_installer/shizuku_apk_installer.dart';

class FleetProfileResult {
  final bool success;
  final int appliedCount;
  final int skippedCount;
  final List<String> errors;
  final String message;

  FleetProfileResult({
    required this.success,
    required this.appliedCount,
    required this.skippedCount,
    required this.errors,
    required this.message,
  });
}

class FleetProfileApplier {
  static void _setInstallMethod(SettingsProvider sp, dynamic v) =>
      sp.installerMode = v as String;
  static void _setExternalInstallerPackage(SettingsProvider sp, dynamic v) =>
      sp.externalInstallerPackage = v as String?;
  static void _setExternalInstallerComponent(SettingsProvider sp, dynamic v) =>
      sp.externalInstallerComponent = v as String?;
  static void _setUseSystemFont(SettingsProvider sp, dynamic v) =>
      sp.useSystemFont = v as bool;
  static void _setTheme(SettingsProvider sp, dynamic v) =>
      sp.theme = ThemeSettings.values[(v as num).toInt()];
  static void _setThemeColor(SettingsProvider sp, dynamic v) =>
      sp.themeColor = Color((v as num).toInt());
  static void _setColourSchemeMode(SettingsProvider sp, dynamic v) =>
      sp.colourSchemeMode = ColourSchemeMode.values[(v as num).toInt()];
  static void _setUseBlackTheme(SettingsProvider sp, dynamic v) =>
      sp.useBlackTheme = v as bool;
  static void _setUpdateInterval(SettingsProvider sp, dynamic v) =>
      sp.updateInterval = (v as num).toInt();
  static void _setCheckOnStart(SettingsProvider sp, dynamic v) =>
      sp.checkOnStart = v as bool;
  static void _setSortColumn(SettingsProvider sp, dynamic v) =>
      sp.sortColumn = SortColumnSettings.values[(v as num).toInt()];
  static void _setSortOrder(SettingsProvider sp, dynamic v) =>
      sp.sortOrder = SortOrderSettings.values[(v as num).toInt()];
  static void _setShowAppWebpage(SettingsProvider sp, dynamic v) =>
      sp.showAppWebpage = v as bool;
  static void _setPinUpdates(SettingsProvider sp, dynamic v) =>
      sp.pinUpdates = v as bool;
  static void _setBuryNonInstalled(SettingsProvider sp, dynamic v) =>
      sp.buryNonInstalled = v as bool;
  static void _setGroupBy(SettingsProvider sp, dynamic v) =>
      sp.groupBy = v as String;
  static void _setHideTrackOnlyWarning(SettingsProvider sp, dynamic v) =>
      sp.hideTrackOnlyWarning = v as bool;
  static void _setHideAPKOriginWarning(SettingsProvider sp, dynamic v) =>
      sp.hideAPKOriginWarning = v as bool;
  static void _setShowAppDowngradeError(SettingsProvider sp, dynamic v) =>
      sp.showAppDowngradeError = v as bool;
  static void _setTactileFeedbackEnabled(SettingsProvider sp, dynamic v) =>
      sp.tactileFeedbackEnabled = v as bool;
  static void _setIncludePrereleasesByDefault(SettingsProvider sp, dynamic v) =>
      sp.includePrereleasesByDefault = v as bool;
  static void _setRemoveOnExternalUninstall(SettingsProvider sp, dynamic v) =>
      sp.removeOnExternalUninstall = v as bool;
  static void _setCheckUpdateOnDetailPage(SettingsProvider sp, dynamic v) =>
      sp.checkUpdateOnDetailPage = v as bool;
  static void _setDisablePageTransitions(SettingsProvider sp, dynamic v) =>
      sp.disablePageTransitions = v as bool;
  static void _setReversePageTransitions(SettingsProvider sp, dynamic v) =>
      sp.reversePageTransitions = v as bool;
  static void _setEnableBackgroundUpdates(SettingsProvider sp, dynamic v) =>
      sp.enableBackgroundUpdates = v as bool;
  static void _setBgUpdatesOnWiFiOnly(SettingsProvider sp, dynamic v) =>
      sp.bgUpdatesOnWiFiOnly = v as bool;
  static void _setBgUpdatesWhileChargingOnly(SettingsProvider sp, dynamic v) =>
      sp.bgUpdatesWhileChargingOnly = v as bool;
  static void _setHighlightTouchTargets(SettingsProvider sp, dynamic v) =>
      sp.highlightTouchTargets = v as bool;
  static void _setDisableSwipeActions(SettingsProvider sp, dynamic v) =>
      sp.disableSwipeActions = v as bool;
  static void _setAlwaysUsePhoneLayout(SettingsProvider sp, dynamic v) =>
      sp.alwaysUsePhoneLayout = v as bool;
  static void _setAutoExportOnChanges(SettingsProvider sp, dynamic v) =>
      sp.autoExportOnChanges = v as bool;
  static void _setExportSettings(SettingsProvider sp, dynamic v) =>
      sp.exportSettings = (v as num).toInt();
  static void _setOnlyCheckInstalledOrTrackOnlyApps(
          SettingsProvider sp, dynamic v) =>
      sp.onlyCheckInstalledOrTrackOnlyApps = v as bool;
  static void _setParallelDownloads(SettingsProvider sp, dynamic v) =>
      sp.parallelDownloads = v as bool;
  static void _setSearchDeselected(SettingsProvider sp, dynamic v) =>
      sp.searchDeselected = (v as List).cast<String>();
  static void _setShowActionBannerForUpdateOnly(SettingsProvider sp, dynamic v) =>
      sp.showActionBannerForUpdateOnly = v as bool;
  static void _setBeforeNewInstallsShareToAppVerifier(
          SettingsProvider sp, dynamic v) =>
      sp.beforeNewInstallsShareToAppVerifier = v as bool;
  static void _setShizukuPretendToBeGooglePlay(SettingsProvider sp, dynamic v) =>
      sp.shizukuPretendToBeGooglePlay = v as bool;
  static void _setCategories(SettingsProvider sp, dynamic v,
          {AppsProvider? appsProvider}) =>
      sp.setCategories(
          (v as Map).map((k, v) => MapEntry(k.toString(), (v as num).toInt())),
          appsProvider: appsProvider);
  static void _setForcedLocale(SettingsProvider sp, dynamic v) =>
      sp.forcedLocale = (v as String?) != null && (v as String).isNotEmpty
          ? _tryParseLocale(v)
          : null;

  static Locale? _tryParseLocale(String? localeString) {
    if (localeString == null) return null;
    final split = localeString.split('-');
    if (split.length == 3) {
      return Locale.fromSubtags(
        languageCode: split[0],
        scriptCode: split[1],
        countryCode: split[2],
      );
    }
    if (split.length == 2) {
      return Locale(split[0], split[1]);
    }
    if (split.isNotEmpty) {
      return Locale(split[0]);
    }
    return null;
  }

  static final Map<String, Map<String, dynamic>> _valueAliases = {
    'theme': {'system': 0, 'light': 1, 'dark': 2},
    'colourSchemeMode': {
      'standard': 0,
      'vibrant': 1,
      'expressive': 2,
      'materialYou': 3,
    },
    'sortColumn': {
      'added': 0,
      'nameAuthor': 1,
      'authorName': 2,
      'releaseDate': 3,
    },
    'sortOrder': {'ascending': 0, 'descending': 1},
    'exportSettings': {'disabled': 0, 'enabled': 1, 'overwrite': 2},
  };

  static dynamic _resolveValue(String key, dynamic value) {
    if (value is! String) return value;
    final aliases = _valueAliases[key];
    if (aliases == null) return value;
    return aliases[value] ?? value;
  }

  static final Map<String, void Function(SettingsProvider, dynamic)>
      _setterMap = {
    'installMethod': _setInstallMethod,
    'installerMode': _setInstallMethod,
    'externalInstallerPackage': _setExternalInstallerPackage,
    'externalInstallerComponent': _setExternalInstallerComponent,
    'useSystemFont': _setUseSystemFont,
    'theme': _setTheme,
    'themeColor': _setThemeColor,
    'colourSchemeMode': _setColourSchemeMode,
    'useBlackTheme': _setUseBlackTheme,
    'updateInterval': _setUpdateInterval,
    'checkOnStart': _setCheckOnStart,
    'sortColumn': _setSortColumn,
    'sortOrder': _setSortOrder,
    'showAppWebpage': _setShowAppWebpage,
    'pinUpdates': _setPinUpdates,
    'buryNonInstalled': _setBuryNonInstalled,
    'groupBy': _setGroupBy,
    'hideTrackOnlyWarning': _setHideTrackOnlyWarning,
    'hideAPKOriginWarning': _setHideAPKOriginWarning,
    'showAppDowngradeError': _setShowAppDowngradeError,
    'tactileFeedbackEnabled': _setTactileFeedbackEnabled,
    'includePrereleasesByDefault': _setIncludePrereleasesByDefault,
    'removeOnExternalUninstall': _setRemoveOnExternalUninstall,
    'checkUpdateOnDetailPage': _setCheckUpdateOnDetailPage,
    'disablePageTransitions': _setDisablePageTransitions,
    'reversePageTransitions': _setReversePageTransitions,
    'enableBackgroundUpdates': _setEnableBackgroundUpdates,
    'bgUpdatesOnWiFiOnly': _setBgUpdatesOnWiFiOnly,
    'bgUpdatesWhileChargingOnly': _setBgUpdatesWhileChargingOnly,
    'highlightTouchTargets': _setHighlightTouchTargets,
    'disableSwipeActions': _setDisableSwipeActions,
    'alwaysUsePhoneLayout': _setAlwaysUsePhoneLayout,
    'autoExportOnChanges': _setAutoExportOnChanges,
    'exportSettings': _setExportSettings,
    'onlyCheckInstalledOrTrackOnlyApps': _setOnlyCheckInstalledOrTrackOnlyApps,
    'parallelDownloads': _setParallelDownloads,
    'searchDeselected': _setSearchDeselected,
    'forcedLocale': _setForcedLocale,
    'showActionBannerForUpdateOnly': _setShowActionBannerForUpdateOnly,
    'beforeNewInstallsShareToAppVerifier':
        _setBeforeNewInstallsShareToAppVerifier,
    'shizukuPretendToBeGooglePlay': _setShizukuPretendToBeGooglePlay,
  };

  /// Whether the profile has a `_meta.grantFirst` flag set.
  static bool _grantFirst(Map<String, dynamic> profile) {
    final meta = profile['_meta'];
    if (meta is! Map) return false;
    return meta['grantFirst'] == true;
  }

  static Future<FleetProfileResult> applyJson(
    SettingsProvider sp, {
    required String json,
    AppsProvider? appsProvider,
  }) async {
    try {
      final parsed = jsonDecode(json);
      if (parsed is! Map) {
        return FleetProfileResult(
          success: false,
          appliedCount: 0,
          skippedCount: 0,
          errors: ['Profile must be a JSON object'],
          message: 'Invalid profile format',
        );
      }
      return await applyMap(sp, parsed.cast<String, dynamic>(),
          appsProvider: appsProvider);
    } catch (e) {
      return FleetProfileResult(
        success: false,
        appliedCount: 0,
        skippedCount: 0,
        errors: [e.toString()],
        message: 'Failed to parse JSON: $e',
      );
    }
  }

  static Future<FleetProfileResult> applyMap(
    SettingsProvider sp,
    Map<String, dynamic> profile, {
    AppsProvider? appsProvider,
  }) async {
    final errors = <String>[];
    var applied = 0;
    var skipped = 0;
    final grantFirst = _grantFirst(profile);

    // If grantFirst is requested, try to grant Shizuku permission before
    // setting installMethod. If it fails, skip installMethod.
    bool? shizukuGranted;
    if (grantFirst && profile['installMethod'] == 'shizuku') {
      try {
        final res = await ShizukuApkInstaller().checkPermission();
        shizukuGranted = res?.startsWith('granted') ?? false;
        if (!shizukuGranted) {
          unawaited(
            LogsProvider().add(
              'Fleet profile: grantFirst enabled but Shizuku permission '
                  'not granted ($res); skipping installMethod',
              level: LogLevel.warning,
            ),
          );
        }
      } catch (e) {
        unawaited(
          LogsProvider().add(
            'Fleet profile: grantFirst Shizuku check failed: $e',
            level: LogLevel.warning,
          ),
        );
        shizukuGranted = false;
      }
    }

    final keys = profile.keys.toList();
    for (final alias in keys) {
      if (alias.startsWith('_')) continue;

      final setter = _setterMap[alias];
      if (setter == null) {
        skipped++;
        errors.add('Unknown setting: $alias');
        continue;
      }

      // Skip installMethod if grantFirst was requested but failed.
      if (alias == 'installMethod' || alias == 'installerMode') {
        if (grantFirst && shizukuGranted == false &&
            profile[alias] == 'shizuku') {
          skipped++;
          errors.add('installMethod: Shizuku permission not granted');
          continue;
        }
      }

      try {
        final resolved = _resolveValue(alias, profile[alias]);
        if (alias == 'categories') {
          _setCategories(sp, resolved, appsProvider: appsProvider);
        } else {
          setter(sp, resolved);
        }
        applied++;
      } catch (e) {
        skipped++;
        errors.add('$alias: $e');
      }
    }

    final message = 'Applied $applied settings, skipped $skipped' +
        (errors.isEmpty ? '' : ' (${errors.length} errors)');

    return FleetProfileResult(
      success: errors.isEmpty,
      appliedCount: applied,
      skippedCount: skipped,
      errors: errors,
      message: message,
    );
  }

  static Future<FleetProfileResult> applyFromFile(
    SettingsProvider sp, {
    required String path,
    AppsProvider? appsProvider,
  }) async {
    try {
      final file = File(path);
      if (!file.existsSync()) {
        return FleetProfileResult(
          success: false,
          appliedCount: 0,
          skippedCount: 0,
          errors: ['File not found: $path'],
          message: 'Failed to read profile from $path',
        );
      }
      final json = file.readAsStringSync();
      return await applyJson(sp, json: json, appsProvider: appsProvider);
    } catch (e) {
      return FleetProfileResult(
        success: false,
        appliedCount: 0,
        skippedCount: 0,
        errors: [e.toString()],
        message: 'Failed to read $path: $e',
      );
    }
  }
}
